src/code/z_room.c

```diff
@@ -1,94 +1,579 @@
+    }
+
+    if ((room->enablePosLights != 0) || (MREG(93) != 0)) {
+        gSPSetGeometryMode(POLY_OPA_DISP++, 0x00400000);
+        gSPSetGeometryMode(POLY_XLU_DISP++, 0x00400000);
+    }
 
-void Room_DrawType1Mesh(PlayState* play, Room* room, u32 flags) {
-    RoomMeshType1* mesh = &room->mesh->type1;
-    if (mesh->format == 1) {
-        func_8012DEE8(play, room, flags);
-    } else if (mesh->format == 2) {
-        func_8012E32C(play, room, flags);
+    roomShape = &room->roomShape->cullable;
+    roomShapeCullableEntry = Lib_SegmentedToVirtual(roomShape->entries);
+    insert = linkedEntriesBuffer;
+
+    roomShapeCullableEntries = roomShapeCullableEntry;
+
+    if (play->roomCtx.unk78 < 0) {
+        for (i = 0; i < roomShape->numEntries; i++, roomShapeCullableEntry++) {
+            if (R_ROOM_CULL_DEBUG_MODE != 0) {
+                if (((R_ROOM_CULL_DEBUG_MODE == ROOM_CULL_DEBUG_MODE_UP_TO_TARGET) &&
+                     (i <= R_ROOM_CULL_DEBUG_TARGET)) ||
+                    ((R_ROOM_CULL_DEBUG_MODE == ROOM_CULL_DEBUG_MODE_ONLY_TARGET) && (i == R_ROOM_CULL_DEBUG_TARGET))) {
+                    if (flags & ROOM_DRAW_OPA) {
+                        displayList = roomShapeCullableEntry->opa;
+                        if (displayList != NULL) {
+                            gSPDisplayList(POLY_OPA_DISP++, displayList);
+                        }
+                    }
+
+                    if (flags & ROOM_DRAW_XLU) {
+                        displayList = roomShapeCullableEntry->xlu;
+                        if (displayList != NULL) {
+                            gSPDisplayList(POLY_XLU_DISP++, displayList);
+                        }
+                    }
+                }
+            } else {
+                if (flags & ROOM_DRAW_OPA) {
+                    displayList = roomShapeCullableEntry->opa;
+                    if (displayList != NULL) {
+                        gSPDisplayList(POLY_OPA_DISP++, displayList);
+                    }
+                }
+
+                if (flags & ROOM_DRAW_XLU) {
+                    displayList = roomShapeCullableEntry->xlu;
+                    if (displayList != NULL) {
+                        gSPDisplayList(POLY_XLU_DISP++, displayList);
+                    }
+                }
+            }
+        }
+    } else {
+        f32 var_fa1 = 1.0f / play->unk_187F0.z; // sp54
+        s32 pad5;
+
+        // Pick and sort entries by depth
+        for (i = 0; i < roomShape->numEntries; i++, roomShapeCullableEntry++) {
+
+            // Project the entry position, to get the depth it is at.
+            pos.x = roomShapeCullableEntry->boundsSphereCenter.x;
+            pos.y = roomShapeCullableEntry->boundsSphereCenter.y;
+            pos.z = roomShapeCullableEntry->boundsSphereCenter.z;
+            SkinMatrix_Vec3fMtxFMultXYZ(&play->viewProjectionMtxF, &pos, &projectedPos);
+            // SkinMatrix_Vec3fMtxFMultXYZW(&play->viewProjectionMtxF, &pos, &projectedPos, &projectedW);
```

Comment:
Why is this commented out?

Reply:
Ah, this was here for reference because OoT's equivalent function uses `SkinMatrix_Vec3fMtxFMultXYZW`, I'll remove it